tools/insn_freq_explorer.c

```diff
@@ -0,0 +1,265 @@
+#endif
+#if RV32_HAS(EXT_M)
+    {"mul", 0},       {"mulh", 0},     {"mulhsu", 0},   {"mulhu", 0},
+    {"div", 0},       {"divu", 0},     {"rem", 0},      {"remu", 0},
+#endif
+#if RV32_HAS(EXT_A)
+    {"lrw", 0},       {"scw", 0},      {"amoswapw", 0}, {"amoaddw", 0},
+    {"amoxorw", 0},   {"amoandw", 0},  {"amoorw", 0},   {"amominw", 0},
+    {"amomaxw", 0},   {"amominux", 0}, {"amomaxux", 0},
+#endif
+#if RV32_HAS(EXT_F)
+    {"flw", 0},       {"fsw", 0},      {"fmadds", 0},   {"fmsubs", 0},
+    {"fnmsubs", 0},   {"fnmadds", 0},  {"fadds", 0},    {"fsubs", 0},
+    {"fmuls", 0},     {"fdivs", 0},    {"fsqrts", 0},   {"fsgnjs", 0},
+    {"fsgnjns", 0},   {"fsgnjxs", 0},  {"fmins", 0},    {"fmaxs", 0},
+    {"fcvtws", 0},    {"fcvtwus", 0},  {"fmvxw", 0},    {"feqs", 0},
+    {"flts", 0},      {"fles", 0},     {"fclasss", 0},  {"fcvtsw", 0},
+    {"fcvtswu", 0},   {"fmvmx", 0},
+#endif
+#if RV32_HAS(EXT_C)
+    {"caddi4spn", 0}, {"clw", 0},      {"csw", 0},      {"cnop", 0},
+    {"caddi", 0},     {"cjal", 0},     {"cli", 0},      {"caddi16sp", 0},
+    {"clui", 0},      {"csrli", 0},    {"csrai", 0},    {"candi", 0},
+    {"csub", 0},      {"cxor", 0},     {"cor", 0},      {"cand", 0},
+    {"cj", 0},        {"cbeqz", 0},    {"cbnez", 0},    {"cslli", 0},
+    {"clwsp", 0},     {"cjr", 0},      {"cmv", 0},      {"cebreak", 0},
+    {"cjalr", 0},     {"cadd", 0},     {"cswsp", 0},
+#endif
+    {"unknown", 0}};
+
+static int cmp_dec(const void *a, const void *b)
+{
+    return ((rv_insn_hist_t *) b)->freq - ((rv_insn_hist_t *) a)->freq;
+}
+
+static int cmp_asc(const void *a, const void *b)
+{
+    return ((rv_insn_hist_t *) a)->freq - ((rv_insn_hist_t *) b)->freq;
+}
+
+/* used to adjust the length of histogram bar */
+static unsigned short get_win_max_col()
+{
+    struct winsize terminal_size;
+    ioctl(STDOUT_FILENO, TIOCGWINSZ, &terminal_size);
+
+    return terminal_size.ws_col;
+}
+
+/*      No.     instruction     percentages [frequency] histogram */
+static const char *fmt = "%3d. %-10s%5.2f%% [%-10zu] %s\n";
+
+/* get columns used by the fmt string */
+static unsigned short get_used_col()
+{
+    unsigned short used_col = 0;
+
+    used_col += 3;  /* width = 3 */
+    used_col += 1;  /* . */
+    used_col += 1;  /* single space */
+    used_col += 10; /* width = 10 */
+    used_col += 5;  /* width = 5 */
+    used_col += 1;  /* % */
+    used_col += 1;  /* single space */
+    used_col += 1;  /* [ */
+    used_col += 10; /* width = 10 */
+    used_col += 1;  /* ] */
+    used_col += 1;  /* single space */
+
+    return used_col;
+}
+
+static char *gen_hist_bar(char *hist_bar,
+                          size_t insn_freq,
+                          size_t max_insn_freq,
+                          unsigned short max_col,
+                          unsigned short used_col)
+{
+    int hist_bar_len = insn_freq * (max_col - used_col) / max_insn_freq;
+    for (int i = 0; i < hist_bar_len; i++) {
+        hist_bar[i] = '*';
```

Comment:
Use the text rendering technique as `_mi_make_bar` (See https://github.com/microsoft/mimalloc/pull/529/files ).